app/src/main/java/com/github/se/eduverse/repository/TodoRepository.kt

```diff
@@ -0,0 +1,143 @@
+   */
+  open fun getActualTodos(
+      userId: String,
+      onSuccess: (List<Todo>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath)
+        .whereEqualTo("ownerId", userId)
+        .whereEqualTo("status", "ACTUAL")
+        .get()
+        .addOnSuccessListener { result ->
+          val todos = result.documents.mapNotNull { document -> decodeTodo(document) }
+          onSuccess(todos)
+        }
+        .addOnFailureListener(onFailure)
+  }
+
+  /**
+   * Get the list of done todos from the database
+   *
+   * @param onSuccess code executed if the todos are successfully accessed
+   * @param onFailure code executed if the todos can't be accessed
+   */
+  open fun getDoneTodos(
+      userId: String,
+      onSuccess: (List<Todo>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath)
+        .whereEqualTo("ownerId", userId)
+        .whereEqualTo("status", "DONE")
+        .get()
+        .addOnSuccessListener { result ->
+          val todos = result.documents.mapNotNull { document -> decodeTodo(document) }
+          onSuccess(todos)
+        }
+        .addOnFailureListener(onFailure)
+  }
+
+  /**
+   * Add a new todo to the database
+   *
+   * @param todo the todo to add
+   * @param onSuccess code executed if the todo is successfully added
+   * @param onFailure code executed if the todo can't be added
+   */
+  open fun addNewTodo(todo: Todo, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    db.collection(collectionPath)
+        .document(todo.uid)
+        .set(todo)
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { onFailure(it) }
+  }
+
+  /**
+   * Update a todo in the database
+   *
+   * @param todo the todo to update
+   * @param onSuccess code executed if the todo is successfully updated
+   * @param onFailure code executed if the todo can't be updated
+   */
+  open fun updateTodo(todo: Todo, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    db.collection(collectionPath)
+        .document(todo.uid)
+        .set(todo)
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { onFailure(it) }
+  }
+
+  /**
+   * Delete a todo from the database
+   *
+   * @param todoId the id of the todo to delete
+   * @param onSuccess code executed if the todo is successfully deleted
+   * @param onFailure code executed if the todo can't be deleted
+   */
+  open fun deleteTodoById(todoId: String, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    db.collection(collectionPath)
+        .document(todoId)
+        .delete()
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { onFailure(it) }
+  }
+
+  /**
+   * Decodes a Firestore document into a ToDo object.
+   *
+   * @param document The Firestore document to decode.
+   * @return The ToDo object.
+   */
+  private fun decodeTodo(document: DocumentSnapshot): Todo? {
```

Comment:
Make sure to handle any null values in the result set, as missing fields could cause decodeTodo to return null and potentially result in an incomplete list.